math/tangent.go

```diff
@@ -1,132 +1,114 @@
-// Output range: all real numbers
-// Notes:
-//		Tan(NaN)    -> NaN
-//		Tan(+/-Inf) -> NaN
-func Tan(z *decimal.Big, theta *decimal.Big) *decimal.Big {
-	//here we use the formula
-	//tan(x) = sign(x)*sqrt(1/cos(x)^2-1)
+// Tan returns the tangent, in radians, of x.
+//
+// Range:
+//     Input: -pi/2 <= x <= pi/2
+//     Output: all real numbers
+//
+// Special cases:
+//     Tan(NaN) = NaN
+//     Tan(±Inf) = NaN
+func Tan(z, x *decimal.Big) *decimal.Big {
+	if z.CheckNaNs(x, nil) {
+		return z
+	}
 
-	if theta.IsInf(0) || theta.IsNaN(0) {
+	if x.IsInf(0) {
 		z.Context.Conditions |= decimal.InvalidOperation
 		return z.SetNaN(false)
 	}
 
-	calculatingPrecision := z.Context.Precision + defaultExtraPrecision
-	x, ok := prepareTangentInput(calculatingPrecision, theta)
+	ctx := decimal.Context{Precision: precision(z) + defaultExtraPrecision}
+	x0, ok := prepTan(z, x, ctx)
 	if !ok {
 		z.Context.Conditions |= decimal.InvalidOperation
 		return z.SetNaN(false)
 	}
 
-	// tangent has an asymptote at pi/2 and we'll need more precision as we get closer
-	// the reason we need it is as we approach pi/2 is due to the squaring portion,
-	// it will cause the small value of Cosine to be come extremely small.
-	// we COULD fix it by simply doubling the precision, however,
-	// when the precision gets larger it will be a significant impact
-	// to performance, instead we'll only add the extra precision when we need it
-	// by using the difference to see how much extra precision we need
-	// we'll speed things up by only using a quick compare to see if we need
-	// to do a deeper inspection
-	tmpCalculation := calculatingPrecision
-	if decimal.New(14, 1).CmpAbs(theta) < 0 {
-		piOver2 := Pi(decimal.WithPrecision(calculatingPrecision))
-		piOver2.Quo(piOver2, two)
-		var dd *decimal.Big
-		if theta.Signbit() {
-			dd = decimal.WithPrecision(calculatingPrecision).Add(theta, piOver2)
+	ctx.Precision++
+
+	// tan(x) = sign(x)*sqrt(1/cos(x)^2-1)
+
+	// tangent has an asymptote at pi/2 and we'll need more precision as we get
+	// closer the reason we need it is as we approach pi/2 is due to the squaring
+	// portion, it will cause the small value of Cosine to be come extremely
+	// small. we COULD fix it by simply doubling the precision, however, when
+	// the precision gets larger it will be a significant impact to performance,
+	// instead we'll only add the extra precision when we need it by using the
+	// difference to see how much extra precision we need we'll speed things up
+	// by only using a quick compare to see if we need to do a deeper inspection.
+	tctx := ctx
+	var tmp decimal.Big
+	if x.CmpAbs(onePtFour) >= 0 {
+		if x.Signbit() {
+			ctx.Add(&tmp, x, pi2(&tmp, ctx))
 		} else {
-			dd = decimal.WithPrecision(calculatingPrecision).Sub(theta, piOver2)
+			ctx.Sub(&tmp, x, pi2(&tmp, ctx))
 		}
-		tmpCalculation = calculatingPrecision + dd.Scale() - dd.Precision()
+		tctx.Precision += ctx.Precision + tmp.Scale() - tmp.Precision()
```

Comment:
`tctx.Precision += ctx.Precision + tmp.Scale() - tmp.Precision()`  your more than doubling the precision by using the `+=` 
should just be `=`

Reply:
nice catch.